src/serdey.rs

```diff
@@ -19,31 +19,27 @@
 // #[cfg(feature = "serde")]
 // use serde::de::Visitor;
 
-
 #[cfg(feature = "serde")]
 macro_rules! serde_boilerplate { ($t:ty) => {
-impl ::serde::Serialize for $t {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: ::serde::Serializer {
-        serializer.serialize_bytes(&self.to_bytes()[..])
+impl ::serde_crate::Serialize for $t {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: ::serde_crate::Serializer {
+        let bytes = &self.to_bytes()[..];
+        ::serde_bytes::Bytes::new(bytes).serialize(serializer)
     }
 }
 
-impl<'d> ::serde::Deserialize<'d> for $t {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: ::serde::Deserializer<'d> {
-        struct MyVisitor;
-
-        impl<'d> ::serde::de::Visitor<'d> for MyVisitor {
-            type Value = $t;
-
-            fn expecting(&self, formatter: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
-                formatter.write_str(Self::Value::DESCRIPTION)
-            }
-
-            fn visit_bytes<E>(self, bytes: &[u8]) -> Result<$t, E> where E: ::serde::de::Error {
-                Self::Value::from_bytes(bytes).map_err(crate::errors::serde_error_from_signature_error)
+impl<'d> ::serde_crate::Deserialize<'d> for $t {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: ::serde_crate::Deserializer<'d> {
+        cfg_if::cfg_if!{
+            if #[cfg(any(feature = "alloc", feature = "std"))] {
+                let bytes = <::serde_bytes::ByteBuf>::deserialize(deserializer)?;
+            } else {
+                let bytes = <&::serde_bytes::Bytes>::deserialize(deserializer)?;
```

Comment:
oh cute, what's going on here?

Reply:
if we are using the `std` or the `alloc` feature during deserialization we should switch to `ByteBuf` since it [handles more cases](https://github.com/serde-rs/bytes/blob/master/src/bytebuf.rs#L196-L244):

- `visit_seq`
- `visit_byte_buf`
- `visit_string`
- `visit_bytes`
- `visit_str`

`Bytes` [handles](https://github.com/serde-rs/bytes/blob/master/src/bytes.rs#L166-L186):
- `visit_borrowed_bytes`
- `visit_borrowed_str`

But after poking around `serde_bytes` a bit I noticed `Cow<'_, [u8]>` is supported to deserialize into, which could slightly be more efficient